modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageCompactionTriggerConfiguration.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.metastorage.impl;
+
+import static org.apache.ignite.internal.util.CompletableFutures.nullCompletedFuture;
+
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.apache.ignite.internal.configuration.SystemDistributedConfiguration;
+import org.apache.ignite.internal.configuration.SystemDistributedView;
+import org.apache.ignite.internal.configuration.SystemPropertyView;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.jetbrains.annotations.Nullable;
+
+/** Configuration for metastorage compaction triggering based on distributed system properties. */
+public class MetaStorageCompactionTriggerConfiguration {
+    private static final IgniteLogger LOG = Loggers.forClass(MetaStorageCompactionTriggerConfiguration.class);
+
+    /**
+     * System property that defines compaction start interval (in milliseconds).
+     *
+     * <p>Default value is {@link #INTERVAL_DEFAULT_VALUE}.</p>
+     */
+    public static final String INTERVAL_SYSTEM_PROPERTY_NAME = "metastorageCompactionInterval";
+
+    /**
+     * System property that defines compaction data availability time (in milliseconds).
+     *
+     * <p>Default value is {@link #DATA_AVAILABILITY_TIME_DEFAULT_VALUE}.</p>
+     */
+    public static final String DATA_AVAILABILITY_TIME_SYSTEM_PROPERTY_NAME = "metastorageCompactionDataAvailabilityTime";
+
+    /** Default value for the {@link #INTERVAL_SYSTEM_PROPERTY_NAME}. */
+    // TODO: IGNITE-23280 Make default 1 minute
+    public static final long INTERVAL_DEFAULT_VALUE = Long.MAX_VALUE;
+
+    /** Default value for the {@link #DATA_AVAILABILITY_TIME_SYSTEM_PROPERTY_NAME}. */
+    // TODO: IGNITE-23280 Make default 1 hour
```

Comment:
When doing MG disaster recovery there is a scenario:

1. Old leader A (the only MG voting member) crashes and cannot start again
2. User repairs MG on remaining nodes
3. Later, old ex-leader from item 1 is migrated to the repaired cluster.

To successfully join, A needs to check its Metastorage checksum against the new leader's checksums. If the default for MS data availability is just 1 hour and more than 1 hour passed between items 2 and 3, the new leader could have already compacted out the latest MS revision which A has, so A will not be able to join.

Could this default be increased?

Reply:
Or maybe use other 'checksum availabilityy time' which is significantly larger than the 'data availability time'?

Reply:
This value is taken randomly, I just thought that this is more or less a sufficient value.

We can increase it, or wait for real bad cases and empirically understand what value is needed.

Also, this value will be subtracted from the current safe time, if during the repair the safe time does not move, then everything will be ok.

Reply:
>Also, this value will be subtracted from the current safe time, if during the repair the safe time does not move, then everything will be ok.

As soon as new leader comes to life, it will update its SafeTime to the current clock time and compact out everything that is older than 1 hour. This doesn't seem to change anything in the provided scenario.

Could we choose 24 hours as a default? Seems to be sufficient.

Reply:
In my opinion, a lot of garbage can accumulate in the metastorage during this time. 
Why do you think that the repair can take more than an hour? Why not more than a day?
The meaning of this property was put on the PDS and no one left a comment.

Reply:
A network partition that is a couple of hours long is a lot more probable than a network partition of 1 day long. By setting this to 1 day, we solve a lot of possible repair problems, and we cannot solve them all as we have compaction.

> The meaning of this property was put on the PDS and no one left a comment.
At that time no one thought about this, now we have a scenario.

Let's not make this a blocker for merging this, but let's file a separate issue if you don't want to change it in this PR.

Reply:
Created https://issues.apache.org/jira/browse/IGNITE-23573